server/iframe.go

```diff
@@ -46,30 +53,175 @@ func (a *API) iFrame(w http.ResponseWriter, _ *http.Request) {
 	}
 }
 
-var iFrameHTML = `<!DOCTYPE html>
-<html>
-<head>
-  <meta charset="UTF-8">
-  <title>Mattermost DevSecOps</title>
-  <meta name="viewport" content="width=device-width, height=device-height, initial-scale=1.0">
-  <script 
-    src="https://res.cdn.office.net/teams-js/2.34.0/js/MicrosoftTeams.min.js"
-    integrity="sha384-brW9AazbKR2dYw2DucGgWCCcmrm2oBFV4HQidyuyZRI/TnAkmOOnTARSTdps3Hwt"
-    crossorigin="anonymous"
-  ></script>
-</head>
-<body>
-    <iframe
-        style="position: absolute; top: 0; left: 0; width: 100%; height: 100%; border: none;"
-        src="{{SITE_URL}}" title="Mattermost DevSecOps" onload="notifyAppLoaded()">
-    </iframe>
-  <script>
-    // Initialize the Microsoft Teams SDK
-    microsoftTeams.app.initialize(["{{SITE_URL}}"]);
-
-    function notifyAppLoaded() {
-      microsoftTeams.app.notifySuccess();
-    }
-  </script>
-</body>
-</html>`
+// formatTemplate formats the iFrame HTML template with the site URL and plugin ID
+func (a *API) formatTemplate(template string) (string, error) {
+	config := a.p.API.GetConfig()
+	siteURL := *config.ServiceSettings.SiteURL
+	if siteURL == "" {
+		return "", fmt.Errorf("ServiceSettings.SiteURL cannot be empty for MS Teams iFrame")
+	}
+
+	html := strings.ReplaceAll(template, "{{SITE_URL}}", siteURL)
+	html = strings.ReplaceAll(html, "{{PLUGIN_ID}}", manifest.Id)
+	html = strings.ReplaceAll(html, "{{TENANT_ID}}", a.p.getConfiguration().TenantID)
+	return html, nil
+}
+
+// authenticate expects a Microsoft Entra ID in the Authorization header, and uses that
+// to authenticate to the corresponding user in Mattermost, if it exists.
+func (a *API) authenticate(w http.ResponseWriter, r *http.Request) {
+	var logger logrus.FieldLogger
+	logger = logrus.StandardLogger()
+
+	// If the user is already logged in, redirect to the home page.
+	// TODO: Refactor the user properties setup to a function and call it from here if the user is already logged in
+	// just in case the user logs in from a tabApp in a browser.
+	if r.Header.Get("Mattermost-User-ID") != "" {
+		logger = logger.WithField("user_id", r.Header.Get("Mattermost-User-ID"))
+		logger.Info("Skipping authentication, user already logged in")
+		http.Redirect(w, r, "/", http.StatusSeeOther)
+		return
+	}
+
+	config := a.p.apiClient.Configuration.GetConfig()
+
+	enableDeveloper := config.ServiceSettings.EnableDeveloper
+
+	// Ideally we'd accept the token via an Authorization header, but for now get it from the query sring.
```

Comment:
```suggestion
	// Ideally we'd accept the token via an Authorization header, but for now get it from the query string.
```

Reply:
I wonder if having the token in a query string will pass security.  The token will appear in proxy logs etc.  What is the life-time of a typical token here?

Reply:
@fmartingr's original approach (before we nested twice) was to exchange this via `window.postMessage`. We can certainly go back to that if it's a blocker.